### ISSUE TITLE
Create a Plan of Action and Make Sense of our Pending Changes

### ISSUE THREAD

**Issue by confused-Techie** [pulsar-edit/package-backend#67]
Alright, @Digitalone1 I'd really like to try and get things under control here if possible, since I'll be honest at this point I'm getting myself a little confused.

Currently we have many backlogged PR's to push to prod, and want to make sure we get everything right. Since we have our migrate to new versioning, rolling back those changes, again migrating to new versioning, and many other features, plus their respective changes to the prod DB, and now we have conflicting changes from the GitHub Refactor and the Versioning changes again.

So if you'd be so kind to help me sort this out to make sure we have a flawless update, and everything get's the proper changes it needs.

Since the last time our Prod server instance was actually updated was 3 weeks ago, Jan 3, 2023 5:42PM. Meaning [this](13003a682aef6566f255113feffb4a7e14dd796c) was likely the last commit actually pushed to prod. Which happened right after merging PR #37 

Meaning since then we are missing the following PRs:

* #41 
* #40 
* #39 
  - Includes the creation of the AuthState Table
* #42 
  - This introduced the new Constraints on the versions, which are well aware that those couldn't be pushed to production, and that's why I had stopped pushing changes, since I wasn't sure the changes made in this PR would work, without the changes to the DB also done.
* #44 
* #45 
* #47 
* #48 
* #49 
* #52 
* #43 
* #55 
* #59 
* #53 
* #60 

So looking at this now, it looks like for backlogged changes we just need to ensure to create the AuthState Table Essentially. But prior to pushing these changes, if we can at this point, I'd want to make sure everything would still work in relation to versions handling without the DB changes. But I think previously you said it'd be best to wait until we get the new changes in.

So with the above said, it looks like it'd be best to wait until #64 is complete? Does that sound correct? Which amazingly it's almost there.

So maybe what's best for us to do right now, is prioritize #65 and #66 and whatever else is left to be needed for the versioning changes. Then we can properly check off a list of the DB changes needed to push all this to production.

After which I can go through and rework the GitHub refactor to function with the newest changes you've made so everything is still compatible. 

And once done we could then get #62 and #50 pushed to main and pushed to prod.

---

So to summarize it all up, (And please correct me if wrong) but once your version changes are all complete and merged this would be the priority to finish  while we essentially have a code freeze.

* [ ] - Create AuthTable (Or ensure it exists and everything is as it should be)
* [ ] - Check off all items left on #64 with your [awesome checklist](https://github.com/pulsar-edit/package-backend/issues/64#issuecomment-1399262414) you've already created.
* [ ] - Ensure there are no additional changes needed from #65 and #66 

Again please correct me if wrong, I just want to make sure we have a flawless release. And thanks again, as always for all your fantastic work here. Sorry on my end I got quite busy with everything else on this project and fell behind keeping up here.

**Comment by Digitalone1** [comment]
Oh, I forgot to say that we use `structuredClone` now, so we should have the highest NodeJS version possible (>=17 at least reading the compatibility table on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone)).

**Comment by confused-Techie** [comment]
> Oh, I forgot to say that we use `structuredClone` now, so we should have the highest NodeJS version possible (>=17 at least reading the compatibility table on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone)).

Awesome, so everything is correct other than also getting NodeJS bumped for us.

But super cool. Then exactly as time permits I'll work on getting the versions with proper creation/updated stamps (For duplicated main versions at least) but glad to hear it can still work without it.

Then once we get that all in I'll finish up the Git Refactor to work with any changes that have happened to main. And we should be good. Appreciate you checking this out

**Comment by Digitalone1** [comment]
@confused-Techie This is a recap in one only place of all the changes to be made on the database.

- First update NodeJS to latest version available on GAE

- [ ] Create new `authstate` table :
https://github.com/pulsar-edit/package-backend/blob/main/scripts/database/create_authstate_table.sql#L1-L5:
```sql
CREATE TABLE authstate (
    id UUID DEFAULT GEN_RANDOM_UUID() PRIMARY KEY,
    keycode VARCHAR(256) NOT NULL UNIQUE,
    created TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP
);
```

- [ ] Create a cleanup cronjob to be run regularly (every 4-6 hours, 4-6 times a day)
https://github.com/pulsar-edit/package-backend/blob/main/scripts/database/create_authstate_table.sql#L9-L10
```sql
DELETE FROM authstate
WHERE created < CURRENT_TIMESTAMP - INTERVAL '10 minutes';
```

- Before pushing #62, apply changes in 
https://github.com/pulsar-edit/package-backend/pull/62#issue-1550034934

- [ ] Drop NOT NULL constraint on `names.pointer`
```sql
ALTER TABLE names ALTER COLUMN pointer DROP NOT NULL;
``` 

- [ ] Retrieve the foreign key constraint name in `names` table
```sql
SELECT conrelid::regclass AS table_name, 
       conname AS foreign_key, 
       pg_get_constraintdef(oid) 
FROM   pg_constraint 
WHERE  contype = 'f' and conrelid::regclass::text = 'names' 
AND    connamespace = 'public'::regnamespace   
ORDER  BY conrelid::regclass::text, contype DESC;
```

- [ ] After the retrieve of the name of the constraint, drop it
```sql
ALTER TABLE names DROP CONSTRAINT replace_here_the_constraint_name;
```

- [ ] Make a new foreign key constraint using the policy of setting NULL on deletion, this way we avoid **names supply chain attack** 
```sql
ALTER TABLE names 
    ADD CONSTRAINT package_names_fkey 
    FOREIGN KEY (pointer) REFERENCES packages(pointer) ON DELETE SET NULL;
```

- Before pushing #65 and #66, apply changes in
https://github.com/pulsar-edit/package-backend/issues/64#issuecomment-1399262414

- [ ] Add new `created` and `updated` column to `versions`
```sql
ALTER TABLE versions
    ADD COLUMN created TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP;
ALTER TABLE versions
    ADD COLUMN updated TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP;
``` 

- [ ] Add update timestamp trigger for `versions`
```sql
CREATE TRIGGER trigger_now_on_updated_versions
    BEFORE UPDATE ON versions
    FOR EACH ROW
EXECUTE PROCEDURE now_on_updated_package();
```

- [ ] Add `deleted` column to `versions`
```sql
ALTER TABLE versions
    ADD COLUMN deleted BOOLEAN NOT NULL DEFAULT FALSE;
```

**Comment by confused-Techie** [comment]
@Digitalone1 Thanks a ton for this!

**Comment by confused-Techie** [comment]
Alright it's time to merge all of these and make the needed changes!

Just gonna go ahead and add all the PR's being merged here simple for the sake of simplicity when, after these are all merged I'll go ahead and add a version onto our backend, writing the changelog.

* #62 
* #65 
* #68 
* #66 
* #69 
* #72 
* #73 
* #75 

With that said, everything is merged, and I'll get started on making the changes needed for the DB

**Comment by confused-Techie** [comment]
Alright @Digitalone1 all the listed changes for the DB have been made here. So looks like all that's left is to go ahead and push these changes to prod!

**Comment by Digitalone1** [comment]
@confused-Techie This issue can be closed.

**Comment by confused-Techie** [comment]
Alright, last change keeping this issue open was the implementation of cleaning up our auth-state table, which has been resolved in [`pulsar-edit/package-frontend#60`](https://github.com/pulsar-edit/package-frontend/pull/60).
